### PR TITLE
Make RestateCloudEnvironments read-only properties public

### PR DESCRIPTION
**Pull request by pcholakov**
Inspired by @michael-k's question in #51 about making the environment URL helper functions public.

### CODE REVIEW

**Comment by pcholakov** [comment]
This doesn't directly address the question from the original PR: I'm not at all opposed to making the helper functions public but I'm just wondering if there's a better API for this that's more discoverable. Is there something we could improve in `RestateCloudEnvironment` that would avoid having to fall back to `RestateEnvironment.fromAttributes()`?

**Comment by michael-k** [comment]
Hm, afaik the `readonly` class attributes are already public as this is the default.

Here's what we do.

We have 1 stack per AWS account that creates the RestateCloudEnvironment.

<details><summary>code</summary>

```ts
import * as cdk from 'aws-cdk-lib'
import * as iam from 'aws-cdk-lib/aws-iam'
import { Secret } from 'aws-cdk-lib/aws-secretsmanager'
import { Construct } from 'constructs'
import restateCdk from '@restatedev/restate-cdk'
import type { EnvironmentId } from '@restatedev/restate-cdk'

const { RestateCloudEnvironment } = restateCdk

export class RestateBootstrapStack extends cdk.Stack {
	constructor (scope: Construct, id: string, props?: cdk.StackProps) {
		super(scope, id, props)

		const apiKey = Secret.fromSecretCompleteArn(this, 'RestateCloudApiKey',
			process.env.RESTATE_API_KEY_SECRET_ARN!
		)

		const environment = new RestateCloudEnvironment(this, 'RestateCloudEnv', {
			environmentId: process.env.RESTATE_ENV_ID! as EnvironmentId,
			apiKey,
		})

		environment.invokerRole.attachInlinePolicy(new iam.Policy(this, 'RestateInvokerPolicy', {
			statements: [
				new iam.PolicyStatement({
					effect: iam.Effect.ALLOW,
					actions: ['lambda:InvokeFunction'],
					resources: [this.formatArn({
						partition: cdk.Aws.PARTITION,
						account: cdk.Aws.ACCOUNT_ID,
						region: cdk.Aws.REGION,
						service: 'lambda',
						resource: 'function:*'
					})],
					conditions: {
						StringEquals: {
							'aws:ResourceTag/RestateInvokeable': ['true'],
						},
					},
				})
			]
		}))

		new cdk.CfnOutput(this, 'InvokerRoleArn', {
			value: environment.invokerRole.roleArn,
			description: 'ARN of IAM role used by restate to invoke Lambda functions',
			exportName: 'RestateInvokerRoleArn',
		})
		new cdk.CfnOutput(this, 'ApiKey', {
			value: apiKey.secretArn,
			description: 'ARN of the SecretsManager secret holding the Restate Cloud API Key',
			exportName: 'RestateCloudApiKey',
		})
	}
}
```

</details>


Here's what one of our other stacks does:

<details><summary>code</summary>

```ts
		const deployer = new ServiceDeployer(this, 'ServiceDeployer', {
			logGroup: new logs.LogGroup(this, 'DeploymentLogs', {
				retention: logs.RetentionDays.ONE_MONTH,
				removalPolicy: cdk.RemovalPolicy.DESTROY,
			}),
		})

		const invokerRole = iam.Role.fromRoleArn(this, 'InvokerRole',
			cdk.Fn.importValue('RestateInvokerRoleArn')
		)
		const authToken = secrets.Secret.fromSecretCompleteArn(this, 'RestateCloudApiKey',
			cdk.Fn.importValue('RestateCloudApiKey')
		)

		const restateEnv = RestateEnvironment.fromAttributes({
			invokerRole,
			adminUrl: adminEndpoint('us', process.env.RESTATE_ENV_ID! as EnvironmentId),
			authToken,
		})

		const function = new lambda.Function(this, …)

		deployer.deployService('workflow', function.currentVersion, restateEnv, {
			skipInvokeFunctionGrant: true
		})

// (…)

type RestateCloudRegion = 'us';

function adminEndpoint (region: RestateCloudRegion, environmentId: EnvironmentId): string {
	const bareEnvId = environmentId.replace(/^env_/, '')
	return `https://${bareEnvId}.env.${region}.restate.cloud:9070`
}
```

</details>

There's no real reason why adminUrl isn't a CFN export like the the role ARN or the API key.

The other stack don't create their own RestateCloudEnvironment because I saw issues with IAM's eventual consistency in every initial deployment.  (And because there's no reason to create a new role for every service.  There's a quota on the number of role. :sweat_smile: )

At the moment RestateCloudEnvironment always creates an `invokerRole`.  Adding it to `RestateCloudEnvironmentProps` would have work for us as well.

**Comment by pcholakov** [comment]
Thanks for the feedback, @michael-k!

> Hm, afaik the readonly class attributes are already public as this is the default.

You are, of course, quite right! Too much context switching ;-)

> We have 1 stack per AWS account that creates the RestateCloudEnvironment.
> ...
> The other stack don't create their own RestateCloudEnvironment because I saw issues with IAM's eventual consistency in every initial deployment. (And because there's no reason to create a new role for every service. There's a quota on the number of roles. 😅 )


This makes a lot of sense. I've hit deployment issues because of IAM propagation delay too, it's very annoying! Ok, so I just pushed a fresh update to the PR which proposes a new `RestateCloudEnvironment.fromAttributes` method. With this change, you'll be able to update your workflow stack to use this approach:

```ts
const restateEnv = RestateCloudEnvironment.fromAttributes({
  invokerRole,
  environmentId: process.env.RESTATE_ENV_ID! as EnvironmentId,
  region: 'eu',
  authToken,
})
```

Thoughts on this approach?

File: lib/restate-constructs/restate-cloud-environment.ts [hunk 1]
```diff
@@ -60,13 +61,36 @@ export class RestateCloudEnvironment extends Construct implements IRestateEnviro
...
     this.adminUrl = adminEndpoint(this.region, props.environmentId);
     this.ingressUrl = ingressEndpoint(this.region, props.environmentId);
   }
 
+  /**
+   * Creates a reference to an existing Restate Cloud environment. Unlike instantiating the construct, this variant does
+   * not attempt to create an invoker role, but still returns an Environment object which can be used to deploy
+   * services.
+   *
+   * @param props environment properties - only `environmentId` and `authToken` are required
```
Comment (michael-k) [inline lib/restate-constructs/restate-cloud-environment.ts:77]:
```suggestion
   * @param props environment properties - only `environmentId` and `apiKey` are required
```

**Review by michael-k — COMMENTED** [review]
Probably a bit unexpected that `RestateCloudEnvironment.fromAttributes()` returns a `RestateEnvironment` with slightly different attributes than `RestateCloudEnvironment`.  But CDK itself also does that with their lookup methods, eg. `ApplicationLoadBalancer.fromLookup()` returns a `LookedUpApplicationLoadBalancer` which has some restrictions.

I can work with `RestateCloudEnvironment.fromAttributes()` :)

**Comment by pcholakov** [comment]
@michael-k I released this as `@restatedev/restate-cdk@0.0.0-SNAPSHOT-20241217085406` - further feedback appreciated! In particular, let me know if you'd like to use this _just_ for obtaining the relevant URLs without deploying any services. We can make the `invokerRole` optional if so.

**Pull request merged**